src/windows.rs

```diff
@@ -393,3 +397,66 @@ pub fn lang() -> impl Iterator<Item = String> {
 
     LangIter { array, index }
 }
+
+/// The processor architecture of the installed operating system.
+// The variants listed below are only platforms where Rust toolchain is
+// supported, to see the full list of architectures supported by Windows, see
+// https://docs.rs/winsafe/latest/winsafe/co/struct.PROCESSOR_ARCHITECTURE.html
+#[repr(u16)]
+#[derive(Debug)]
+pub enum ProcessorArchitecture {
+    /// x86
+    Intel = 0,
+    /// Arm
+    Arm = 5,
+    /// x64 (AMD or Intel)
+    Amd64 = 9,
+    /// Arm64
+    Arm64 = 12,
+    /// ?
+    Arm32OnWin64 = 13,
+    /// Unknown architecture
+    Unknown = 0xffff,
+}
+
+#[repr(C)]
+pub struct SystemInfo {
+    pub w_processor_architecture: ProcessorArchitecture,
+    w_reserved: u16,
+    pub dw_page_size: u32,
+    pub lp_minimum_application_address: *mut c_void,
+    pub lp_maximum_application_address: *mut c_void,
+    pub dw_active_processor_mask: usize,
+    pub dw_number_of_processors: u32,
+    pub dw_processor_type: u32,
+    pub dw_allocation_granularity: u32,
+    pub w_processor_level: u16,
+    pub w_processor_revision: u16,
+}
+
+impl Default for SystemInfo {
+    fn default() -> Self {
+        unsafe { std::mem::zeroed() }
+    }
+}
+
+extern "C" {
+    #[link_name = "GetNativeSystemInfo"]
+    fn get_native_system_info(buf: *mut SystemInfo);
+}
+
+pub fn arch() -> Arch {
+    let mut buf = SystemInfo::default();
+    unsafe { get_native_system_info(&mut buf as *mut SystemInfo) };
+
+    match buf.w_processor_architecture {
+        ProcessorArchitecture::Intel => Arch::I686,
+        ProcessorArchitecture::Arm => Arch::Arm,
+        ProcessorArchitecture::Arm64 => Arch::Aarch64,
+        ProcessorArchitecture::Amd64 => Arch::X86_64,
+        ProcessorArchitecture::Arm32OnWin64 => Arch::Arm,
+        ProcessorArchitecture::Unknown => {
+            Arch::Unknown(format!("{:?}", buf.w_processor_architecture))
+        }
```

Comment:
Care should be taken here to check the catch-all case.  Windows could add to the enum at any time, which means checking the enum exhaustively is undefined behavior.  I would consider doing something like [this](https://github.com/ardaku/whoami/blob/8445e54135db829fe9d51b882fe69c6e87598f16/src/windows.rs#L305-L310) rather than trying to define the enum.

Reply:
Changes to something like this:

```rust
// windows.rs 

pub fn arch() -> Arch {
    let mut buf = SystemInfo::default();
    unsafe { get_native_system_info(&mut buf as *mut SystemInfo) };


    // The variants listed below are only platforms where Rust toolchain is
    // supported, to see the full list of architectures supported by Windows, see
    // https://docs.rs/winsafe/latest/winsafe/co/struct.PROCESSOR_ARCHITECTURE.html
    match buf.w_processor_architecture {
        0 => Arch::I686,
        5 => Arch::Arm,
        12 => Arch::Aarch64,
        9 => Arch::X64,
        13 => Arch::Arm,
        unknown_arch_number => Arch::Unknown(
            format!("Unknown Arch Number: {}", unknown_arch_number)
        ),
    }
}
```